syncher/syncserver.go

```diff
@@ -0,0 +1,194 @@
+	}
+	defer conn.Close()
+	client := proto.NewPolicySyncClient(conn)
+	stream, err := client.Sync(cxt, &proto.SyncRequest{})
+	if err != nil {
+		log.Fatalf("failed to Sync with server: %v", err)
+	}
+	for {
+		update, err := stream.Recv()
+		if err != nil {
+			log.Fatalf("connection to Policy Sync server broken: %v", err)
+		}
+		log.WithFields(log.Fields{"proto": update.String()}).Debug("Received sync API Update")
+		store.Write(func(ps *policystore.PolicyStore) { processUpdate(ps, update) })
+	}
+	// Note that as written, this function will never return. It only ends when the connection is torn down, which
+	// terminates the entire program.
+}
+
+// Update the PolicyStore with the information passed over the Sync API.
+func processUpdate(store *policystore.PolicyStore, update *proto.ToDataplane) {
+	switch payload := update.Payload.(type) {
+	case *proto.ToDataplane_InSync:
+		processInSync(store, payload.InSync)
+	case *proto.ToDataplane_IpsetUpdate:
+		processIPSetUpdate(store, payload.IpsetUpdate)
+	case *proto.ToDataplane_IpsetDeltaUpdate:
+		processIPSetDeltaUpdate(store, payload.IpsetDeltaUpdate)
+	case *proto.ToDataplane_IpsetRemove:
+		processIPSetRemove(store, payload.IpsetRemove)
+	case *proto.ToDataplane_ActiveProfileUpdate:
+		processActiveProfileUpdate(store, payload.ActiveProfileUpdate)
+	case *proto.ToDataplane_ActiveProfileRemove:
+		processActiveProfileRemove(store, payload.ActiveProfileRemove)
+	case *proto.ToDataplane_ActivePolicyUpdate:
+		processActivePolicyUpdate(store, payload.ActivePolicyUpdate)
+	case *proto.ToDataplane_ActivePolicyRemove:
+		processActivePolicyRemove(store, payload.ActivePolicyRemove)
+	case *proto.ToDataplane_WorkloadEndpointUpdate:
+		processWorkloadEndpointUpdate(store, payload.WorkloadEndpointUpdate)
+	case *proto.ToDataplane_WorkloadEndpointRemove:
+		processWorkloadEndpointRemove(store, payload.WorkloadEndpointRemove)
+	default:
+		panic(fmt.Sprintf("unknown payload %v", update.String()))
+	}
+}
+
+func processInSync(store *policystore.PolicyStore, inSync *proto.InSync) {
+	// TODO (spikecurtis): disallow requests until policy is synced
+	log.Debug("Processing InSync")
+	return
+}
+
+func processIPSetUpdate(store *policystore.PolicyStore, update *proto.IPSetUpdate) {
+	log.WithFields(log.Fields{
+		"id": update.Id,
+	}).Debug("Processing IPSetUpdate")
+
+	// IPSetUpdate replaces the existing set.
+	s := policystore.NewIPSet(update.Type)
+	store.IPSetByID[update.Id] = s
+	for _, addr := range update.Members {
+		s.AddString(addr)
+	}
+}
+
+func processIPSetDeltaUpdate(store *policystore.PolicyStore, update *proto.IPSetDeltaUpdate) {
+	log.WithFields(log.Fields{
+		"id": update.Id,
+	}).Debug("Processing IPSetDeltaUpdate")
+	s := store.IPSetByID[update.Id]
+	if s == nil {
+		log.Errorf("Unknown IPSet id: %v", update.Id)
+		panic("unknown IPSet id")
+	}
+	for _, addr := range update.AddedMembers {
+		s.AddString(addr)
+	}
+	for _, addr := range update.RemovedMembers {
+		s.RemoveString(addr)
+	}
```

Comment:
Does the semantics of `DeltaUpdate` ensure that the `AddedMembers` and `RemovedMemebers` are disjoint.

Reply:
It better!